Add support for M1/M2 mac in PytorchModelHandlerTensor

bfcrampton opened the pull request:
Add support for M1/M2 mac

Add support for M1/M2 mac by allowing `mps` pytorch device type.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [X] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
   - N/A
 - [X] Update `CHANGES.md` with noteworthy changes.
   - N/A
 - [X] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).
   - N/A

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

damccorm reviewed (commented):
Thanks, this looks like a helpful change. I think we need a few more things for completeness though.

1) Could you please update `_load_model` to add a check like https://github.com/apache/beam/blob/416c3adb25580fad5ced31c69b98b7cd8681f407/sdks/python/apache_beam/ml/inference/pytorch_inference.py#L89 for mps with a clear warning/fallback to CPU if mps isn't available (similar to the code snippet here - https://github.com/apache/beam/blob/416c3adb25580fad5ced31c69b98b7cd8681f407/sdks/python/apache_beam/ml/inference/pytorch_inference.py#L89)
2) Could you please update the pydoc for each of the modified functions to call out that mps is an allowed device type?
3) We should probably prefer `MPS` to `mps` to stay consistent with our `GPU` parameter. Alternately, we can just make both case insensitive (this would probably be for the best anyways).

damccorm commented:
next action author

damccorm commented:
waiting on author

bfcrampton commented:
Thanks for the review!

Addressed the comments. On #3 I decided to make everything case insensitive as well as support specifying `cuda` or `CUDA` directly. This is a minor nit that I have to workaround when using this library, everywhere else a `device` would be either `cpu`, `cuda`, or `mps`.

Lmk what you think, happy to just leave the case insensitivity part.

damccorm commented on sdks/python/apache_beam/ml/inference/pytorch_inference.py:
```diff
@@ -91,6 +91,12 @@ def _load_model(
         "Model handler specified a 'GPU' device, but GPUs are not available. "
         "Switching to CPU.")
     device = torch.device('cpu')
+  if device == torch.device('mps') and not (torch.backend.mps.is_available() and
```
Ah, it also looks like `torch.device('mps')` causes older versions of pytorch to throw. To maintain compat with older versions, could we wrap this in a try except? It should be fine to silently swallow the exception since we'll throw below if someone tries to set the device to mps and the pytorch version doesn't support it.

This is why the test checks are failing
